beamer/check/commands.py

```diff
@@ -0,0 +1,152 @@
+)
+@click.argument("proof-source", type=ChainIdParam())
+@click.argument("proof-target", type=ChainIdParam(), nargs=-1, required=True)
+def initiate_l1_invalidations(
+    keystore_file: Path,
+    password: str,
+    rpc_file: Path,
+    artifacts_dir: Path,
+    abi_dir: Path,
+    output: Path,
+    proof_source: ChainId,
+    proof_target: tuple[ChainId],
+) -> None:  # pylint: disable=unused-argument
+    """Initiate L1 invalidations from chain PROOF_SOURCE to each PROOF_TARGET.
+    PROOF_SOURCE and PROOF_TARGET are chain IDs."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    account = beamer.util.account_from_keyfile(keystore_file, password)
+    log.info("Loaded keystore file", address=account.address)
+
+    abi_manager = ABIManager(abi_dir)
+    deployment = beamer.artifacts.load(artifacts_dir, proof_source)
+
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+    url = rpc_info[proof_source]
+    w3 = make_web3(url, account)
+    assert w3.eth.chain_id == proof_source
+    log.info("Connected to RPC", url=url)
+
+    fill_manager = obtain_contract(w3, abi_manager, deployment, "FillManager")
+    request_manager = obtain_contract(w3, abi_manager, deployment, "RequestManager")
+
+    data = dict(proof_source=proof_source)
+    for chain_id in proof_target:
+        # Mark check-related fill IDs so that we can distiguish them
+        # on chain from other invalidations, if the need arises.
+        fill_id = HexBytes(b"beamer check " + random.randbytes(19))
+        assert len(fill_id) == 32
+
+        # The nonce is only 96 bits.
+        nonce = random.randint(0, 2**96 - 1)
+
+        func = fill_manager.functions.invalidateFill(
+            chain_id, ADDRESS_ZERO, ADDRESS_ZERO, TokenAmount(0), nonce, fill_id
+        )
+        receipt = _transact(func)
+        block = w3.eth.get_block(receipt["blockNumber"])
+
+        # Compute the timestamp of message finalization.
+        finality_period = request_manager.functions.chains(chain_id).call()[0]
+        timestamp = datetime.utcfromtimestamp(block["timestamp"]) + timedelta(
+            seconds=finality_period
+        )
+        request_id = create_request_id(
+            chain_id,
+            proof_source,
+            to_checksum_address(ADDRESS_ZERO),
+            to_checksum_address(ADDRESS_ZERO),
+            TokenAmount(0),
+            nonce,
+        )
+        data[str(chain_id)] = dict(
+            request_id=request_id.hex(),
+            fill_id=fill_id.hex(),
+            txhash=receipt.transactionHash.hex(),  # type: ignore
+            finalization_timestamp=timestamp.isoformat(timespec="seconds"),
+        )
+
+        # Write the output immediately so that we have at least partial data
+        # in case an error occurs during invalidation for the next chain.
+        with output.open("wt") as f:
+            json.dump(data, f, indent=4)
+        log.info("Invalidation info stored", path=str(output))
```

Comment:
Maybe we can at least add chain_id as log here like proof_source -> chain_id or something like that so we dont see duplicates.

Reply:
We now have more info in the logs.